File: uva_10226.py
# -*- coding: utf-8 -*-
"""
Authors: Hung-Hsin Chen <chenhh@par.cse.nsysu.edu.tw>
License: GPL v2
status: on judge
difficulty: 1

https://uva.onlinejudge.org/external/102/10226.pdf
"""
import bisect


def main():
    cases = int(input())

    case_cnt = 0
    species = []
    while True:
        try:
            tree = input().strip()

            if len(tree) > 0:
                bisect.insort_right(species, tree)
            else:
                if case_cnt >= 1 and case_cnt <= cases:
                    n_tree = len(species)
                    if n_tree == 0:
                        print ("")
                    else:
                        sdx = 0
                        # print ("{} {}: {}".format(case_cnt, n_tree, species))
                        for tdx, tree in enumerate(species):
                            if ((tdx >=1 and species[tdx-1] != tree)):
                                # print (tdx, sdx)
                                percent = (tdx-sdx)/n_tree * 100
                                print ("{} {:.4f}".format(
                                    species[sdx], percent ))
                                sdx = tdx
                        # print last set
                        percent = (n_tree-sdx)/n_tree * 100
                        print("{} {:.4f}".format(species[sdx], percent))

                    if case_cnt != cases:
                        print("")

                case_cnt += 1
                species = []

            if case_cnt > cases:
                break
        except (EOFError):
            break

if __name__ == '__main__':
    main()

File: uva_10226.cpp
/*
Authors: Hung-Hsin Chen <chenhh@par.cse.nsysu.edu.tw>
License: GPL v2
status: TLE
difficulty: 1
https://uva.onlinejudge.org/external/102/10226.pdf
*/

#include <iostream>
#include <iomanip>
#include <string>
#include <map>
using namespace std;

int main()
{
    string dummy;
    int number_of_cases;
    cin >> number_of_cases;
    getline(cin, dummy);
    getline(cin, dummy);
    for (int idx = 0; idx < number_of_cases; idx++)
    {
        map<string, int> tree_count;
        int count = 0;
        while (true) {
            string tree;
            getline(cin, tree);
            if (tree.length() == 0) {
                break;
            }
            map<string, int>::iterator probe = tree_count.find(tree);
            if (probe == tree_count.end())
                tree_count.insert(pair<string, int>(tree, 1));
            else
                tree_count[tree]++;
            count++;
        }
        double factor = 100.0 / count;
        cout << setprecision(4) << fixed;
        map<string, int>::iterator iter;
        for ( iter= tree_count.begin(); iter != tree_count.end(); iter++){
            cout << iter->first << " " <<iter->second * factor << endl;
        }

        if (idx != number_of_cases - 1)
            cout << endl;
    }
    return 0;
}
